ui: decouple html base; allow development with short builds

dhartunian opened the pull request:
Previously, a few small issues prevented the UI development process from
making independent choices from CRDB. First, the base HTML template
contained data available in a global variable. Second, the "short" build
would break this template and keep us from developing on the UI when
we're just tweaking server endpoints.

This PR solves both problems. First, the templated variables are removed
and retrieved a startup in `index.tsx`. The React bootstrap waits for
the JSON response which is saved in a global variable by the JS code
instead of the backend. An attempt was made to remove the need for this
global variable but that changes quite a bit of code. In particular, all
of our links to docs pages rely on the global availability of the
cluster version number. Changing this would require that all docs link
generation functions take the version number as input instead of being
statically generated.

Second, the webpack dev server proxy will now attack a header to its
requests allowing us to notice this on the CRDB server and short circuit
the response on short builds to render the base HTML template as if it
was a "full" build. When the webpack dev server is running and serving
dev assets from memory, this lets us interact with DB Console on the
3000 port and develop using the hot reload workflow. This will shorten
dev feedback loops by making builds faster when backend endpoints are
being tweaked.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/78216)
<!-- Reviewable:end -->

dhartunian committed:
ui: decouple html base; allow development with short builds

Previously, a few small issues prevented the UI development process from
making independent choices from CRDB. First, the base HTML template
contained data available in a global variable. Second, the "short" build
would break this template and keep us from developing on the UI when
we're just tweaking server endpoints.

This PR solves both problems. First, the templated variables are removed
and retrieved a startup in `index.tsx`. The React bootstrap waits for
the JSON response which is saved in a global variable by the JS code
instead of the backend. An attempt was made to remove the need for this
global variable but that changes quite a bit of code. In particular, all
of our links to docs pages rely on the global availability of the
cluster version number. Changing this would require that all docs link
generation functions take the version number as input instead of being
statically generated.

If the initial data fails to load from the server, DB Console will show
a critical alert in the top right corner and invite the customer to
refresh the page.

Second, the webpack dev server proxy will now attack a header to its
requests allowing us to notice this on the CRDB server and short circuit
the response on short builds to render the base HTML template as if it
was a "full" build. When the webpack dev server is running and serving
dev assets from memory, this lets us interact with DB Console on the
3000 port and develop using the hot reload workflow. This will shorten
dev feedback loops by making builds faster when backend endpoints are
being tweaked.

Release note: None

matthewtodd reviewed (approved):
:tada: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a couple of questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-:-N3ZrLAc28ZufiVIFGzG:bxjzis)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @sjbarag)

___
*[`pkg/ui/workspaces/db-console/src/index.tsx` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3ZoR9L2cFm_MLv-IVn:-N3ZoR9M34wWr0MYAhY3:bhxyt01) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/index.tsx#L22)):*
> ```XML
>   fetchDataFromServer,
>   setDataFromServer,
> } from "oss/src/util/dataFromServer";
> ```

IIUC we want to strip the leading `oss/` here, yes?
___
*[`pkg/ui/workspaces/db-console/src/util/dataFromServer.ts` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3ZqIQ3DFIRyYVBRYTI:-N3ZqIQ3DFIRyYVBRYTJ:b1p0oul) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/util/dataFromServer.ts#L11)):*
> ```XML
> // licenses/APL.txt.
> 
> import { AdminUIState } from "oss/src/redux/state";
> ```

And this one?
___
*[`pkg/ui/workspaces/db-console/src/util/dataFromServer.ts` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3Zqw3O5RJZOB9-62ig:-N3Zqw3O5RJZOB9-62ih:b-x3leu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/util/dataFromServer.ts#L46)):*
> ```XML
> 
> export const dataFromServerSelector = (s: AdminUIState) =>
>   s.cachedData.dataFromServer;
> ```

Delete this selector? (Looks unused, not sure that `s.cachedData.dataFromServer` is a thing in this revision.)


<!-- Sent from Reviewable.io -->

sjbarag reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78216)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @matthewtodd, and @sjbarag)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_1VJB4xrgb5DGFPna:-N3_1VJB4xrgb5DGFPnb:bt4n8uh) (no related file):*
Looks like a great start @dhartunian!  There's a few small edges want to shave off before this gets released though — mostly the panic, but I wouldn't mind having a separate route for the JSON config response :wink: 

Thanks for tackling both the backend and UI sides of this — it's going to be a *massive* quality-of-life improvement :heart: 

___
___
*[`pkg/ui/ui.go` line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3Zvmca6miUmZrZ_9Ow:-N3Zvmca6miUmZrZ_9Ox:b-yryoyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L131)):*
> ```Go
> 
> 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
> 		log.Infof(r.Context(), "%v", r)
> ```

nit: I assume this is a debugging log that we'll remove before this ships :)
___
*[`pkg/ui/ui.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3ZwUSS8Bb9XXvO9WI7:-N3ZwUSS8Bb9XXvO9WI8:b-wqwknb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L144)):*
> ```Go
> 			OIDCButtonText:       oidcConf.ButtonText,
> 		}
> 		if r.Header.Get("content-type") == "application/json" {
> ```

Using `Content-Type` to determine the response type feels a bit strange to me.  As a request header, it seems like Content-Type is intended to indicate what type of payload the client is sending, but the `GET` here doesn't include anything.  I imagine `Accept: application/json` is closer to what we're looking for here?  But since the responses are semantically different (one is the webpage, the other is a config blob for the webpage), leaning on `Accept` to change the meaning of the response doesn't feel right.  Maybe `/uiconfig` or something like that?



_Code quote:_
```Go
		if r.Header.Get("content-type") == "application/json" {
			argBytes, err := json.Marshal(args)
			if err != nil {
				http.Error(w, err.Error(), 500)
				return
			}
			w.Write(argBytes)
			return
		}
```
___
*[`pkg/ui/ui.go` line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_0P9KBLQgoHMI4Go_:-N3_0P9KBLQgoHMI4Goa:b-h9hfbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L154)):*
> ```Go
> 		}
> 
> 		if !HaveUI && r.Header.Get("X-Crdb-Development") == "" {
> ```

nit: looks like the `X-` prefix for headers [has been deprecated](https://www.rfc-editor.org/rfc/rfc6648), which would make this just `Crdb-Development`.  I don't really mind either way, but :rainbow: :star: the more you know :shrug: 
___
*[`pkg/ui/ui.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_01xLCkPIhWTMoTRo:-N3_01xLCkPIhWTMoTRp:b-tv2gjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L162)):*
> ```Go
> 			fileHandlerChain.ServeHTTP(w, r)
> 			return
> 		}
> ```

Watch out!  This panics when requesting a non-index.html file with the `X-Crdb-Development` set.

`curl --verbose -H 'X-Crdb-Development: 1' localhost:8080/foo`

```
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   |   GOROOT/src/runtime/panic.go:1038
*   | runtime.panicmem
*   |   GOROOT/src/runtime/panic.go:221
*   | runtime.sigpanic
*   |   GOROOT/src/runtime/signal_unix.go:735
*   | net/http.ioFS.Open
*   |   GOROOT/src/net/http/fs.go:760
*   | net/http.serveFile
*   |   GOROOT/src/net/http/fs.go:597
*   | net/http.(*fileHandler).ServeHTTP
*   |   GOROOT/src/net/http/fs.go:848
*   | github.com/cockroachdb/cockroach/pkg/util/httputil.EtagHandler.func1
*   |   github.com/cockroachdb/cockroach/pkg/util/httputil/handlers.go:63
*   | net/http.HandlerFunc.ServeHTTP
*   |   GOROOT/src/net/http/server.go:2047
*   | github.com/cockroachdb/cockroach/pkg/ui.Handler.func1
*   |   github.com/cockroachdb/cockroach/pkg/ui/ui.go:160
*   | net/http.HandlerFunc.ServeHTTP
*   |   GOROOT/src/net/http/server.go:2047
*   | github.com/cockroachdb/cockroach/pkg/server.(*authenticationMux).ServeHTTP
*   |   github.com/cockroachdb/cockroach/pkg/server/authentication.go:582
*   | net/http.(*ServeMux).ServeHTTP
*   |   GOROOT/src/net/http/server.go:2425
*   | github.com/cockroachdb/cockroach/pkg/server.(*httpServer).gzipHandler
*   |   github.com/cockroachdb/cockroach/pkg/server/server_http.go:304
*   | github.com/cockroachdb/cockroach/pkg/server.(*nodeProxy).nodeProxyHandler
*   |   github.com/cockroachdb/cockroach/pkg/server/node_http_router.go:63
*   | github.com/cockroachdb/cockroach/pkg/server.(*httpServer).baseHandler
*   |   github.com/cockroachdb/cockroach/pkg/server/server_http.go:335
*   | net/http.HandlerFunc.ServeHTTP
*   |   GOROOT/src/net/http/server.go:2047
*   | net/http.serverHandler.ServeHTTP
*   |   GOROOT/src/net/http/server.go:2879
*   | net/http.(*conn).serve
*   |   GOROOT/src/net/http/server.go:1930
*   | runtime.goexit
*   |   GOROOT/src/runtime/asm_amd64.s:1581
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
```



_Code quote:_
```Go
		if r.URL.Path != "/" {
			fileHandlerChain.ServeHTTP(w, r)
			return
		}
```
___
*[`pkg/ui/workspaces/db-console/src/index.tsx` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_15GGEatatqCbtL3V:-N3_15GH71YsYUNgN-QU:b-f5xlhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/index.tsx#L24)):*
> ```XML
> } from "oss/src/util/dataFromServer";
> 
> fetchDataFromServer().then(d => {
> ```

:thinking: We'll need to figure out how to handle request failures.  There's already a global `<ErrorBoundary/>` in db-console, so perhaps this calls `setDataFromServer` with an empty string and lets the error boundary render the Sleepy Moon image and text?



_Code quote:_
```XML
fetchDataFromServer().then(d => {
  setDataFromServer(d);

  ReactDOM.render(
    <App history={history} store={store} />,
    document.getElementById("react-layout"),
  );

  store.subscribe(alertDataSync(store));
});
```


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78216)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @matthewtodd, and @sjbarag)

___
*[`pkg/ui/ui.go` line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3Zvmca6miUmZrZ_9Ow:-N3tUByXAI95jXEXGIqL:bmfxgaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L131)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

nit: I assume this is a debugging log that we'll remove before this ships :)
</blockquote></details>

Done :)
___
*[`pkg/ui/ui.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3ZwUSS8Bb9XXvO9WI7:-N3tVe1K4rDS-USGErM6:b-5oqtp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L144)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

Using `Content-Type` to determine the response type feels a bit strange to me.  As a request header, it seems like Content-Type is intended to indicate what type of payload the client is sending, but the `GET` here doesn't include anything.  I imagine `Accept: application/json` is closer to what we're looking for here?  But since the responses are semantically different (one is the webpage, the other is a config blob for the webpage), leaning on `Accept` to change the meaning of the response doesn't feel right.  Maybe `/uiconfig` or something like that?
</blockquote></details>

Agreed. This is mostly the result of trying to quickly hack, not deliberate design. Added a route. Since there's only a single route, I'm keeping the definition in the handler and not adding a router or anything fancy.
___
*[`pkg/ui/ui.go` line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_0P9KBLQgoHMI4Go_:-N3tUdprDIzjSI0_9maS:bh6fsn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L154)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

nit: looks like the `X-` prefix for headers [has been deprecated](https://www.rfc-editor.org/rfc/rfc6648), which would make this just `Crdb-Development`.  I don't really mind either way, but :rainbow: :star: the more you know :shrug: 
</blockquote></details>

lol I def learned this practice *after* the publish date of that RFC. thx for the tip!
___
*[`pkg/ui/ui.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_01xLCkPIhWTMoTRo:-N3uc-Cj3vmUvdCncpZK:b-fdr770) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L162)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

Watch out!  This panics when requesting a non-index.html file with the `X-Crdb-Development` set.

`curl --verbose -H 'X-Crdb-Development: 1' localhost:8080/foo`

```
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   |   GOROOT/src/runtime/panic.go:1038
*   | runtime.panicmem
*   |   GOROOT/src/runtime/panic.go:221
*   | runtime.sigpanic
*   |   GOROOT/src/runtime/signal_unix.go:735
*   | net/http.ioFS.Open
*   |   GOROOT/src/net/http/fs.go:760
*   | net/http.serveFile
*   |   GOROOT/src/net/http/fs.go:597
*   | net/http.(*fileHandler).ServeHTTP
*   |   GOROOT/src/net/http/fs.go:848
*   | github.com/cockroachdb/cockroach/pkg/util/httputil.EtagHandler.func1
*   |   github.com/cockroachdb/cockroach/pkg/util/httputil/handlers.go:63
*   | net/http.HandlerFunc.ServeHTTP
*   |   GOROOT/src/net/http/server.go:2047
*   | github.com/cockroachdb/cockroach/pkg/ui.Handler.func1
*   |   github.com/cockroachdb/cockroach/pkg/ui/ui.go:160
*   | net/http.HandlerFunc.ServeHTTP
*   |   GOROOT/src/net/http/server.go:2047
*   | github.com/cockroachdb/cockroach/pkg/server.(*authenticationMux).ServeHTTP
*   |   github.com/cockroachdb/cockroach/pkg/server/authentication.go:582
*   | net/http.(*ServeMux).ServeHTTP
*   |   GOROOT/src/net/http/server.go:2425
*   | github.com/cockroachdb/cockroach/pkg/server.(*httpServer).gzipHandler
*   |   github.com/cockroachdb/cockroach/pkg/server/server_http.go:304
*   | github.com/cockroachdb/cockroach/pkg/server.(*nodeProxy).nodeProxyHandler
*   |   github.com/cockroachdb/cockroach/pkg/server/node_http_router.go:63
*   | github.com/cockroachdb/cockroach/pkg/server.(*httpServer).baseHandler
*   |   github.com/cockroachdb/cockroach/pkg/server/server_http.go:335
*   | net/http.HandlerFunc.ServeHTTP
*   |   GOROOT/src/net/http/server.go:2047
*   | net/http.serverHandler.ServeHTTP
*   |   GOROOT/src/net/http/server.go:2879
*   | net/http.(*conn).serve
*   |   GOROOT/src/net/http/server.go:1930
*   | runtime.goexit
*   |   GOROOT/src/runtime/asm_amd64.s:1581
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
```
</blockquote></details>

Thanks for catching this. It's confusing because it seems like the current behavior when `HaveUI == false` is to return the index HTML on all routes. I'd like it to 404 on unknown routes which seems more correct, but might leave that out of this PR for clarity.

The modified code will return the index html on *all* Paths if the header is present.
___
*[`pkg/ui/workspaces/db-console/src/index.tsx` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_15GGEatatqCbtL3V:-N3ySBDiDDpX6WW-WphD:b-ckqy9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/index.tsx#L24)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

:thinking: We'll need to figure out how to handle request failures.  There's already a global `<ErrorBoundary/>` in db-console, so perhaps this calls `setDataFromServer` with an empty string and lets the error boundary render the Sleepy Moon image and text?
</blockquote></details>

see what you think of my approach here...feels a bit hacky. Maybe I misunderstood your suggestion.
___
*[`pkg/ui/workspaces/db-console/src/util/dataFromServer.ts` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3Zqw3O5RJZOB9-62ig:-N3tV-Dq5IHqHIoq59JC:br78ivk) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/util/dataFromServer.ts#L46)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Delete this selector? (Looks unused, not sure that `s.cachedData.dataFromServer` is a thing in this revision.)
</blockquote></details>

thanks for flagging this. yeah I was trying to put the data in redux in an earlier iteration and it turned into a mess I could not unravel.


<!-- Sent from Reviewable.io -->

sjbarag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78216)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian and @matthewtodd)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N442_-x8OJziaez0WrK:-N442_-yDYyv_1yw7Nav:b-17t4ku) (no related file):*
Looking great!  Let's make sure the `/uiconfig` failure case is handled gracefully, but otherwise this looks ready to go!

___
___
*[`pkg/ui/ui.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3ZwUSS8Bb9XXvO9WI7:-N44-ZkW2G4z6FTA96ff:b-5txruo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L144)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Agreed. This is mostly the result of trying to quickly hack, not deliberate design. Added a route. Since there's only a single route, I'm keeping the definition in the handler and not adding a router or anything fancy.
</blockquote></details>

Gotcha!  Thanks for the quick fix.  This should be fine :)
___
*[`pkg/ui/ui.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_01xLCkPIhWTMoTRo:-N4407cx1NUDSXkM8g9B:b-s5rq6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/ui.go#L162)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Thanks for catching this. It's confusing because it seems like the current behavior when `HaveUI == false` is to return the index HTML on all routes. I'd like it to 404 on unknown routes which seems more correct, but might leave that out of this PR for clarity.

The modified code will return the index html on *all* Paths if the header is present.
</blockquote></details>

Works for me :)
___
*[`pkg/ui/ui.go` line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N44-lLX4KhRq08A5lbz:-N44-lLX4KhRq08A5lc-:blc6y4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f57a2940be41b4fffc9642c81d33ac178f51cb4/pkg/ui/ui.go#L149)):*
> ```Go
> 			argBytes, err := json.Marshal(args)
> 			if err != nil {
> 				http.Error(w, err.Error(), 500)
> ```

nit: think it's worth adding a `log` line here? I genuinely can't tell.



_Code quote:_
```Go
http.Error(w, err.Error(), 500)
```
___
*[`pkg/ui/workspaces/db-console/src/index.tsx` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_15GGEatatqCbtL3V:-N440lXY22ypaqKE2Hnz:bm80jh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/index.tsx#L24)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

see what you think of my approach here...feels a bit hacky. Maybe I misunderstood your suggestion.
</blockquote></details>

Sorry — I should have clarified!  My thinking was something like:

```ts
    setDataFromServer(
        fetchDataFromServer().catch(() => null);
    );

    ReactDOM.render(
        <App history = {history} store={store}/>,
        document.getElementById("react-layout"),
    );
    store.subscribe(alertDataSync(store));
})();
```

The hope is that if `window.dataFromServer` is malformed, something throws an Error that gets caught by the `<ErrorBoundary/>` that's on every page (https://github.com/cockroachdb/cockroach/blob/0c1ab841f5877c45ec50df10c9956bb295b8095d/pkg/ui/workspaces/db-console/src/views/app/containers/layout/index.tsx#L103).  Perhaps that's not actually what happens in the case of a malformed `window.dataFromServer` though?

Anyway, this would require us to make `dataFromServer` optioanl on `Window`, which is technically true now.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78216)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @matthewtodd, and @sjbarag)

___
*[`pkg/ui/ui.go` line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N44-lLX4KhRq08A5lbz:-N49-ks53Ful9DdzTirr:blqbcjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f57a2940be41b4fffc9642c81d33ac178f51cb4/pkg/ui/ui.go#L149)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

nit: think it's worth adding a `log` line here? I genuinely can't tell.
</blockquote></details>

Sure, why not :) Done.
___
*[`pkg/ui/workspaces/db-console/src/index.tsx` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_15GGEatatqCbtL3V:-N49-nT12JALT3flkFHg:bzdhy49) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/index.tsx#L24)):*
<details><summary><i>Previously, sjbarag (Sean Barag) wrote…</i></summary><blockquote>

Sorry — I should have clarified!  My thinking was something like:

```ts
    setDataFromServer(
        fetchDataFromServer().catch(() => null);
    );

    ReactDOM.render(
        <App history = {history} store={store}/>,
        document.getElementById("react-layout"),
    );
    store.subscribe(alertDataSync(store));
})();
```

The hope is that if `window.dataFromServer` is malformed, something throws an Error that gets caught by the `<ErrorBoundary/>` that's on every page (https://github.com/cockroachdb/cockroach/blob/0c1ab841f5877c45ec50df10c9956bb295b8095d/pkg/ui/workspaces/db-console/src/views/app/containers/layout/index.tsx#L103).  Perhaps that's not actually what happens in the case of a malformed `window.dataFromServer` though?

Anyway, this would require us to make `dataFromServer` optioanl on `Window`, which is technically true now.
</blockquote></details>

Ah I understand. Thanks for clarifying. We fail the request "silently" and expect errors resulting from its absence to hit the Boundary once the render tries to do a thing.

So I gave this a shot and nothing really broke. The compromise I've reached is adding an alert banner that's triggered when `dataFromServer` is an empty object, which is what we return if it's never set. This lets the customer know that a problem occurred but doesn't keep them from using DB Console if they choose to. In the error case, the affected features are: login config stuff (OIDC, etc.), retrieving the currently logged-in user's username to display, constructing valid docs URLs, knowing the ID of the current node you're on. Nothing that's a huge problem.


![Screenshot 2022-06-10 at 12-46-15 Cluster Overview Cockroach Console.png](https://files.reviewableusercontent.io/images/reviewable/986307/OVpGinKBiUA3tYcOd4_IU-vV/Screenshot%202022-06-10%20at%2012-46-15%20Cluster%20Overview%20Cockroach%20Console.png)


<!-- Sent from Reviewable.io -->

dhartunian commented:
friendly ping @sjbarag

sjbarag reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78216)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @matthewtodd, and @sjbarag)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N560ejP8q1hB1a-KMLP:-N560ejP8q1hB1a-KMLQ:b-3cezuy) (no related file):*
Yesssssssssss let's ship this!

___
___
*[`pkg/ui/workspaces/db-console/src/index.tsx` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78216#-N3_15GGEatatqCbtL3V:-N560Aze8IpE5D_c3Hu4:bbjnx2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa5b248e31583114457b957755f0012aec9e92d3/pkg/ui/workspaces/db-console/src/index.tsx#L24)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Ah I understand. Thanks for clarifying. We fail the request "silently" and expect errors resulting from its absence to hit the Boundary once the render tries to do a thing.

So I gave this a shot and nothing really broke. The compromise I've reached is adding an alert banner that's triggered when `dataFromServer` is an empty object, which is what we return if it's never set. This lets the customer know that a problem occurred but doesn't keep them from using DB Console if they choose to. In the error case, the affected features are: login config stuff (OIDC, etc.), retrieving the currently logged-in user's username to display, constructing valid docs URLs, knowing the ID of the current node you're on. Nothing that's a huge problem.


![Screenshot 2022-06-10 at 12-46-15 Cluster Overview Cockroach Console.png](https://files.reviewableusercontent.io/images/reviewable/986307/OVpGinKBiUA3tYcOd4_IU-vV/Screenshot%202022-06-10%20at%2012-46-15%20Cluster%20Overview%20Cockroach%20Console.png)

</blockquote></details>

That's a great compromise!


<!-- Sent from Reviewable.io -->

dhartunian commented:
TFTRs!

bors r=sjbarag,matthewtodd
